woocommerce/woocommerce
# fix: wcadmin react18 createroot print shipping banner

* Related to #46879

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR changes the creation of the React root used to show our React based WooCommerce Shipping promotional banner by switching from using `render` to `createRoot`, enabling React 18 mode behaviour.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

_**NOTE**: The banner we're displaying (see screenshot below) is depending on a few conditions to be true, but the most cumbersome to work around is to install Jetpack and make a connection via your WordPress.com account since that requires your website to be publicly accessible (read: reachable outside of your local network)._

_To simplify the test instructions, since we only care about the banner itself being able to be rendered correctly, I'm assuming you're able to spin up this branch locally and will be able to make a small PHP modification:_

1. Spin up this branch locally.
2. The meta box we're testing is only hooked into the post type version of WC_Order, so we need to make sure HPOS is disabled:
    1. Go to WooCommerce > Settings > Advanced > Features (`/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`)
    2. In `Order data storage`, make sure `WordPress posts storage (legacy)` is enabled.
    3. _Trick: if you're using a Woo installation that already have orders, then you can switch the setting after you've enabled compatibility mode and waited for the orders to be synced_. 
4. Run `pnpm --filter=@woocommerce/plugin-woocommerce build` from root to build the updated React code.
5. Add `return true;` to the top of `ShippingLabelBanner::should_show_meta_box()` in `plugins/woocommerce/src/Internal/Admin/ShippingLabelBanner.php`.
    * _This will force the banner to be rendered and thereby work around the caveat mentioned in the above note._
6. Go to the "Add new order" page (`/wp-admin/post-new.php?post_type=shop_order`)
7. You should now see the banner from the screenshot below.
8. Open your developer console.
9. Verify that you're not seeing any `ReactDOM.render` warning (`Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead`).
    * _Please note that if you see `Warning: You are importing createRoot from "react-dom" which is not supported. You should instead import it from "react-dom/client".` then it's unrelated to this change and it something that was introduces in WordPress 6.6 and is being worked on by the Gutenberg team: https://github.com/WordPress/gutenberg/issues/62923#issuecomment-2199438175_.



### Screenshot

![Screenshot 2024-08-09 at 16 39 25](https://github.com/user-attachments/assets/f23f83d1-ef25-4ea2-a252-1e7a8ee9dc2a)

### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
